main.rb

```diff
@@ -0,0 +1,13 @@
+def self.new_with_session(params, session)  
+    super.tap do |user|
+        if data = session["devise.facebook_data"] && session["devise.facebook_data"]["extra"]["raw_info"]      
+            user.email = data["email"] if user.email.blank?    
+        end  
+    end 
+end 
+def self.from_omniauth(auth)  
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|    
+        user.email = auth.info.email    
+        user.password = Devise.friendly_token[0,20]  
+    end 
+end
```

Comment:
We are in the JavaScript section, therefore please find the JS way of doing this. Personally, it doesn't make sense to use Ruby in a vanilla JS section project. You'll have a chance to work with both in the final graduation Capstone project.